### ISSUE TITLE
strange response to OSC 11 background color query

### ISSUE THREAD

**Issue by dankamongmen** [christianparpart/contour#233]
## Description

Using tip of current contour (Contour Terminal Emulator 0.2.0-unreleased-master-aaf30e1), I'm performing a background color query via `\e]11;?\e\\`. While I cannot find this control sequence in any standards, it does seem implemented by many modern terminals, including XTerm, Alacritty, and Kitty.

Contour clearly interprets the query, as it replies with a X color spec of the form `rgb:r/g/b`, where `r`, `g`, and `b` are 4-digit hexadecimal numbers padded to the left with 0s. Good good. The problem is how this reply is terminated. The other aforementioned terminals all follow the `g` component with an ESC and a ST, but `contour` seems to terminate the reply with a 0x7 BEL.

Maybe this is allowed (i've updated my state machine to tolerate and use it), but it seems an unnecessary deviation from common practice.

**Contour**
`"\33]11;rgb:1D1D/1F1F/2121\7"`
**Kitty**
`"\33]11;rgb:0000/2222/1111\33\\"`
**Alacritty**
`"\33]11;rgb:1d1d/1f1f/2121\33\\"`
**XTerm**
`""\33]11;rgb:3326/3326/3326\33\\"`

So....if you could replace the BEL with a `\e\\`, that would be great.

## Environment

- aaf30e1e4e20686b8d2b9c56f69904ec8d7dc783
- Linux 5.12.11nlb, debian unstable
- ```color_schemes:                                                                                             
    default:                                                                                               
        # Default colors                                                                                   
        default:                                                                                           
            # Default background color (this can be made transparent, see above).                          
            background: '#1d1f21'                                                                          
            # Default foreground text color.                                                               
            foreground: '#d5d8d6'  ```
- `TERM=xterm-256color`

## Steps to Reproduce

```c
write(3, "\33]11;?\33\\\33[=0c\33[>c\33[>q\33P+q544e\33\\\33[?1;3;256S\33[?2;1;0S\33[?1;1;0S\33[c", 63) = 63
read(0, "\33]11;rgb:1D1D/1F1F/2121\7", 8192) = 24
read(0, "\33[>65;200;0c\33[?1;0;256S\33[?1;0;800S\33[?1;0;256S\33[?65;22;29;314;1;4;8c", 8192) = 67
```

**Comment by dankamongmen** [comment]
also, is there any recommended query-based method of identifying `contour`, or am i stuck with heuristics and `TERMINAL_NAME`?

**Comment by dankamongmen** [comment]
btw if what you're doing is legal by some standard, point me at it, and i'll update my out-of-spec bullshit. i figure you have your reasons.

**Comment by christianparpart** [comment]
Ooooooooh you have found an old one. I mean, i am since quite some time terminating such sequences with `\033\\`.

Why i used to do BEL instead of ST is because of xterm. In ctlseqs.txt i did read that it is terminating it all with BEL. Only later i learned that this is an xterm invention and it used to be ST. I would recommend to support BEL but i will change that on my side too.

More on the other questions later.

**Comment by dankamongmen** [comment]
yep, i've already updated to handle it

```c
    case STATE_BGSEMI: // drain string                                                                     
      if(c == '\x07'){ // contour sends this at the end for some unknown reason                            
        if(stash_string(inits)){                                                                           
          return -1;                                                                                       
        }                                                                                                  
        inits->state = STATE_NULL;                                                                         
        break;                                                                                             
      }                                                                                                    
      inits->numeric = c;                                                                                  
      if(ruts_string(inits, STATE_BG1)){                                                                   
        return -1;                                                                                         
      }                                                                                                    
      break;                        
```

https://github.com/dankamongmen/notcurses/issues/1780

**Comment by christianparpart** [comment]
> also, is there any recommended query-based method of identifying `contour`, or am i stuck with heuristics and `TERMINAL_NAME`?

Relying on environment variables is highly unreliable. I do not like that. having `TERMINAL_NAME` was an interim solution. I'd rather really like to reserve an VT sequence  for identification request, so that a string can be returned.

So let's do this:

### Terminal Device Identification Request:

```
CSI ? Ps x
```
| Ps   | Description | Example |
|-------|----------------|--------------|
| 1     | respond with terminal's short name | Contour |
| 2     | respond with terminal's version number | 0.2.0-git-BRANCH-SHA |
| 3     | respond with terminal's name and version number | Contour 0.2.0 |

### Response format

```
DCS Ps ; <response in cleartext> ST
```

`DCS` is `\x1bP`, `Ps` is equal to the request's parameter, and `<response text>` is the textual sequence respective to the request parameter `Ps`. Also, I don't think that needs further encoding, since it's text that is requested, that can (or should) not break the VT stream.
If you think that is viable, I'd go ahead and implement that (in the hope that other TEs do jump on **that** generic solution for device textual identification, too).

### Why CSI req and DCS resp?

Because I saw in the vt100.net specs that there are other request/response VT sequences acting exactly the same way.

### BEL vs ST

ctlseqs.txt does mention that in line 1664 (BEL) and 1666 (ST). ctlseqs.txt is also documenting OSC 11 (and OSC 10 for foreground). So I think this is where the others have gotten that sequence from.

p.s.: check out what `OSC 50` is  doing. (some having it implemented. it's quite fun).

**Comment by dankamongmen** [comment]
heh alright =]. here's what i'm sending for identification queries already, since literally everyone has decided to go their own way on this:

```c
  const char queries[] = CSI_BGQ                                                                           
                         "\x1b[=0c"                                                                        
                         "\x1b[>c"                                                                         
                         "\x1b[>q"                                                                         
                         "\x1bP+q544e\x1b\\"                                                               
                         "\x1b[?1;3;256S"                                                                  
                         "\x1b[?2;1;0S"                                                                    
                         "\x1b[?1;1;0S"                                                                    
                         ESC_DA;     
```

there's an XTVERSION, an XTGETTCAP['TN'], teritary, secondary, and primary device attributes...good times

**Comment by christianparpart** [comment]
`TN` of XTGETTCAP simply returns what you should or might want to have in your TERM env var. I don't think that's good. 

But oh damn. I didn't know about `XTVERSION` (documented in ctlseq.txt, my holy bible). I'll go with that one then of course. (btw, I tried a few terminals, kitty/alacritty/konsole/gnome-terminal, noone responsed to `"\x1b[>q"`, I am sad. But I at least had wezterm responding to it. And since it's documented in ctlseqs.txt already, I'll follow that. (what's xterm.js doing @jerch?)

**Comment by dankamongmen** [comment]
> `TN` of XTGETTCAP simply returns what you should or might want to have in your TERM env var. I don't think that's good.

**except** for the particular terminal that identifies itself this way.

trust me. you know way more about what terminals are supposed to do. having spent the past two weeks sorting through all this with many, many existing terminals, i know the grim realities. =\

**Comment by dankamongmen** [comment]
https://github.com/dankamongmen/notcurses/blob/master/src/lib/input.c#L650-L1103 if you want to know the grutesomeness

**Comment by christianparpart** [comment]
> https://github.com/dankamongmen/notcurses/blob/master/src/lib/input.c#L650-L1103 if you want to know the grutesomeness

apparently your xterm is responding to `\033[>q`. Why not mine. Maybe I am missing something in `.Xresources`? Hmm... Anyways. I'm now also responding to exactly that. I like reusing VT sequences for the common good. :-)

![image](https://user-images.githubusercontent.com/56763/122299854-d0800d00-cefe-11eb-8c8f-ef94124d6e5d.png)

I'm actually quite happy that you tested this. Many thanks for taking the time.

**Comment by whisperity** [comment]
Putting this out here just in case, because I think the whole `TERMINAL_NAME` stuff was my idea #167. To clarify: I'm not against implementing magic commands and sequences for better identification, but I am **firmly against** removing the existing cheap and straightforward solution.

#168 showed us that there is a lot of crap going on when it comes to interfacing with these kinds of stuff from a script. Especially when said script really shouldn't *reach* the underlying terminal emulator, but the logic the script scaffolds depends on the environment it's running in.

    echo -ne '\033[>q'

This is unreadable if you aren't sleeping and waking with terminal emulator specs under your pillow. People who end up reading this code in any script will literally flip their sh\*\* if any bugs pop up. `if [[ $TERMINAL_NAME == "Konsole" ]]` is much more readable, and straight on point.

What's more, the problem with one particular emulator resolving one particular sequence and defining one particular response scheme is that this solution is not mathematically any different from the current situation of environment variables with TE X and TE Y implementing their own magic variables in which their own conjured magic response is present, like Contour using `TERMINAL_NAME` while Konsole uses `KONSOLE_DBUS_SESSION` or whatever.

    echo -ne '\033[>q'

The problem with this line is that it actually has to execute, and it actually has to make the emulator framework produce an output, which actually has to be read, parsed, and understood. This introduces a lot of weak points in the pipeline that can come crashing down on anyone who'd like a "I want Vim / Zsh / Whatever to behave X when Contour and Y when Konsole and Z when Tmux" (which is my specific use-case, I have Konsole/Contour specific Vim settings -- mostly because of key captures and idiosyncracies of Konsole, which is the production TE I'm using --, and I have bare/tmux-specific ZSH bindings, mostly around the `capture` (#168) and `tmux capture-pane` feature).

**Comment by christianparpart** [comment]
I think for shell scripts, the terminal emulator generic environment variable `TERMINAL_NAME` is definitely handy, and I (of course) won't remove that ( :100: ). For native code I think it's better to have something slightly more reliable since env vars can be easier forged then rewriting the VT stream. :)

As far as I am concerned with other terminal emulators, I hope they will all adopt to XTVERSION but also as well as the env var `TERMINAL_NAME`. Both are generic versions, each over a different communication pipeline with each having their own pros and cons. I'll be supporting both and won't revert either of them. :-)

**Comment by jerch** [comment]
> But oh damn. I didn't know about XTVERSION (documented in ctlseq.txt, my holy bible). I'll go with that one then of course. (btw, I tried a few terminals, kitty/alacritty/konsole/gnome-terminal, noone responsed to "\x1b[>q", I am sad. But I at least had wezterm responding to it. And since it's documented in ctlseqs.txt already, I'll follow that. (what's xterm.js doing @jerch?)

Not sure, what you are asking me here, got kinda lost between the topics. In general feature reporting is very weak in xterm.js, it even contains code to alter DA1/2 on purpose, which is a really bad practice and totally unreliable for app side (no clue what the original author was thinking here). I tried to address and discuss that several times, but we kinda got stuck in "if we changed that for good, app XY will break until they fix it". If it was only me, I would have done it regardless these upstream issues, but the wide spread usage makes that really a showstopper. Also see https://github.com/xtermjs/xterm.js/issues/2894.
Not implemented: OSC 11, XTVERSION, XTGETTCAP, DA3
Misaligned on purpose: DA1/2

In general I think the new feature reporting sequence is more than needed to fix most of the shortcomings across all TEs.

About BEL vs ST:
BEL was added by xterm as OSC terminator like 20ys ago, and was the only terminator for OSC in xterm for a long time. Imho Thomas fixed that some releases ago now supporting both for OSC. Imho BEL never worked for DCS. By the spec (ECMA-48) only ST is a valid terminator for any string sequence type (thus OSC, DCS, APC, PM).
